tgui-next/packages/tgui/interfaces/Uplink.js

```diff
@@ -0,0 +1,227 @@
+/* eslint-disable react/prefer-stateless-function */
+import { toFixed } from 'common/math';
+import { decodeHtmlEntities } from 'common/string';
+import { Fragment, Component } from 'inferno';
+import { act } from '../byond';
+import { Box, Button, LabeledList, NumberInput, Section, Tabs, Table, Input } from '../components';
+import { map } from 'common/fp';
+import { createLogger } from '../logging';
+import { classes } from 'common/react';
+
+const logger = createLogger("uplink");
+
+// It's a class because we need to store state in the form of the current hovered item, and current search terms
+export class Uplink extends Component {
+
+  constructor() {
+    super();
+    this.state = {
+      hoveredItem: {},
+      currentSearch: "",
+    };
+  }
+
+  setHoveredItem(hoveredItem) {
+    this.setState({
+      hoveredItem: hoveredItem,
+    });
+  }
+
+  setSearchText(currentSearch) {
+    this.setState({
+      currentSearch: currentSearch,
+    });
+  }
+
+  render() {
+    const { state } = this.props;
+    const { config, data } = state;
+    const { ref } = config;
+    const {
+      compact_mode,
+      lockable,
+      telecrystals,
+      categories = [],
+    } = data;
+    const { hoveredItem, currentSearch } = this.state;
+    const hoveredCost = hoveredItem.cost || 0;
+    return (
+      <Section
+        title={(
+          <Box
+            inline
+            color={telecrystals > 0 ? "good" : "bad"}
+          >
+            {telecrystals} TC
+          </Box>
+        )}
+        buttons={(
+          <Fragment>
+            Search
+            <Input
+              value={currentSearch}
+              onInput={(e, value) => this.setSearchText(value)}
+              ml={1}
+              mr={1}
+            />
+            <Button
+              icon={compact_mode ? "list" : "info"}
+              content={compact_mode ? "Compact" : "Detailed"}
+              onClick={() => act(ref, "compact_toggle")}
+            />
+            {!!lockable && (
+              <Button
+                icon="lock"
+                content="Lock"
+                onClick={() => act(ref, "lock") /* lock */}
+              />
+            )}
+          </Fragment>
+        )}
+      >
+        {currentSearch.length > 0 ? (
+          <table className="Table">
+            {categories.map(category => {
```

Comment:
Ideally, this should've been a flatMap, discarding all category info. And I expected you to extract the list of items into a separate component. Usage of "classes" is also redundant, because all they do is concatenate strings, which you could've done manually, yielding even more performance.

All of this is not critical, but it would certainly be nice.

Reply:
classes usage fixed. I didn't extract the list into a separate component because I was being really lazy and didn't want to have to figure out how to get the child component to interact with the state of the parent one. Same for not flattening the map, originally I was going to have the result divided by sections but it didn't look very good.